From 08bc8fd79eadcd24a63e73f5676a173cfe9f7970 Mon Sep 17 00:00:00 2001
From: Thomas Bruederli <thomas@roundcube.net>
Date: Mon, 31 Oct 2022 21:45:28 +0100
Subject: [PATCH 1/3] Add config option for request uri field (#8738)

This can be used to read a custom header sent by a reverse proxy to resolve the absolute path to Roundcube
---
 config/defaults.inc.php    | 7 +++++++
 program/include/rcmail.php | 5 +++--
 2 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/config/defaults.inc.php b/config/defaults.inc.php
index bc5a582150..9d090b6d51 100644
--- a/config/defaults.inc.php
+++ b/config/defaults.inc.php
@@ -817,6 +817,13 @@
 // Note: Use assets_path to not prevent the browser from caching assets
 $config['use_secure_urls'] = false;
 
+// Specify the $_SERVER field that contains the full path of the original HTTP request.
+// This might be changed when Roundcube runs behind a reverse proxy using a subpath.
+// The reverse proxy config can specify a custom header (e.g. X-Forwarded-Path) containing
+// the path under which Roundcube is exposed to the outside world (e.g. /rcube/).
+// This header vaue is then availbale in PHP with $_SERVER['HTTP_X_FORWARDED_PATH'].
+$config['request_uri_field'] = 'REQUEST_URI';
+
 // Allows to define separate server/path for image/js/css files
 // Warning: If the domain is different cross-domain access to some
 // resources need to be allowed
diff --git a/program/include/rcmail.php b/program/include/rcmail.php
index bb13cdc730..dae5cc374a 100644
--- a/program/include/rcmail.php
+++ b/program/include/rcmail.php
@@ -1154,8 +1154,9 @@ public function url($p, $absolute = false, $full = false, $secure = false)
             $prefix = rtrim($prefix, '/') . '/';
         }
         else {
-            if (isset($_SERVER['REQUEST_URI'])) {
-                $prefix = preg_replace('/[?&].*$/', '', $_SERVER['REQUEST_URI']) ?: './';
+            $server_var = $this->config->get('request_uri_field');
+            if (isset($server_var) &&!empty($_SERVER[$server_var])) {
+                $prefix = preg_replace('/[?&].*$/', '', $_SERVER[$server_var]) ?: './';
             }
             else {
                 $prefix = './';

From 64131d8bd83ef1d55902838fb68dc5c82b06989d Mon Sep 17 00:00:00 2001
From: Thomas Bruederli <thomas@roundcube.net>
Date: Wed, 2 Nov 2022 21:49:57 +0100
Subject: [PATCH 2/3] Add check against the proxy_whitelist option

... before using a HTTP header field value for the request uri composition.

Refactor the rcmail::url() method to also work when composing fully qualified urls.
---
 program/include/rcmail.php            | 26 ++++++++++++++++++--------
 program/lib/Roundcube/rcube_utils.php |  9 ++++++++-
 2 files changed, 26 insertions(+), 9 deletions(-)

diff --git a/program/include/rcmail.php b/program/include/rcmail.php
index dae5cc374a..a3bbc3e2f3 100644
--- a/program/include/rcmail.php
+++ b/program/include/rcmail.php
@@ -1119,7 +1119,11 @@ public function url($p, $absolute = false, $full = false, $secure = false)
         }
 
         $base_path = '';
-        if (!empty($_SERVER['REDIRECT_SCRIPT_URL'])) {
+        $server_var = $this->get_requesr_uri_field();
+        if ($server_var && !empty($_SERVER[$server_var])) {
+            $base_path = preg_replace('/[?&].*$/', '', $_SERVER[$server_var]) ?: './';
+        }
+        else if (!empty($_SERVER['REDIRECT_SCRIPT_URL'])) {
             $base_path = $_SERVER['REDIRECT_SCRIPT_URL'];
         }
         else if (!empty($_SERVER['SCRIPT_NAME'])) {
@@ -1154,18 +1158,24 @@ public function url($p, $absolute = false, $full = false, $secure = false)
             $prefix = rtrim($prefix, '/') . '/';
         }
         else {
-            $server_var = $this->config->get('request_uri_field');
-            if (isset($server_var) &&!empty($_SERVER[$server_var])) {
-                $prefix = preg_replace('/[?&].*$/', '', $_SERVER[$server_var]) ?: './';
-            }
-            else {
-                $prefix = './';
-            }
+            $prefix = $base_path ?: './';
         }
 
         return $prefix . $url;
     }
 
+    /**
+     * Get the 'request_uri_field' config option
+     * with an additional check against the 'proxy_whitelist' config
+     */
+    protected function get_requesr_uri_field() {
+        $server_var = $this->config->get('request_uri_field');
+        if (!empty($server_var) && (strpos($server_var, 'HTTP_') !== 0 || rcube_utils::check_proxy_whitelist_ip())) {
+            return $server_var;
+        }
+        return null;
+    }
+
     /**
      * Function to be executed in script shutdown
      */
diff --git a/program/lib/Roundcube/rcube_utils.php b/program/lib/Roundcube/rcube_utils.php
index 75c227638d..29e8b93ad3 100644
--- a/program/lib/Roundcube/rcube_utils.php
+++ b/program/lib/Roundcube/rcube_utils.php
@@ -673,7 +673,7 @@ public static function https_check($port = null, $use_https = true)
 
         if (!empty($_SERVER['HTTP_X_FORWARDED_PROTO'])
             && strtolower($_SERVER['HTTP_X_FORWARDED_PROTO']) == 'https'
-            && in_array($_SERVER['REMOTE_ADDR'], (array) rcube::get_instance()->config->get('proxy_whitelist', []))
+            && self::check_proxy_whitelist_ip()
         ) {
             return true;
         }
@@ -689,6 +689,13 @@ public static function https_check($port = null, $use_https = true)
         return false;
     }
 
+    /**
+     * Check if the reported REMOTE_ADDR is in the 'proxy_whitelist' config option
+     */
+    public static function check_proxy_whitelist_ip() {
+        return in_array($_SERVER['REMOTE_ADDR'], (array) rcube::get_instance()->config->get('proxy_whitelist', []));
+    }
+
     /**
      * Replaces hostname variables.
      *

From d0d268588f20f8468b104b65b7f6d3394c74cdf2 Mon Sep 17 00:00:00 2001
From: Thomas Bruederli <thomas@roundcube.net>
Date: Sun, 6 Nov 2022 22:48:28 +0100
Subject: [PATCH 3/3] Refactor & fix/adapt tests

---
 program/include/rcmail.php  |  7 ++++---
 tests/Rcmail/OutputHtml.php |  4 ++--
 tests/Rcmail/Rcmail.php     | 14 ++++++++++----
 3 files changed, 16 insertions(+), 9 deletions(-)

diff --git a/program/include/rcmail.php b/program/include/rcmail.php
index a3bbc3e2f3..f4f5e32982 100644
--- a/program/include/rcmail.php
+++ b/program/include/rcmail.php
@@ -1119,9 +1119,9 @@ public function url($p, $absolute = false, $full = false, $secure = false)
         }
 
         $base_path = '';
-        $server_var = $this->get_requesr_uri_field();
+        $server_var = $this->get_request_uri_field();
         if ($server_var && !empty($_SERVER[$server_var])) {
-            $base_path = preg_replace('/[?&].*$/', '', $_SERVER[$server_var]) ?: './';
+            $base_path = preg_replace('/[?&].*$/', '', $_SERVER[$server_var]);
         }
         else if (!empty($_SERVER['REDIRECT_SCRIPT_URL'])) {
             $base_path = $_SERVER['REDIRECT_SCRIPT_URL'];
@@ -1168,7 +1168,8 @@ public function url($p, $absolute = false, $full = false, $secure = false)
      * Get the 'request_uri_field' config option
      * with an additional check against the 'proxy_whitelist' config
      */
-    protected function get_requesr_uri_field() {
+    protected function get_request_uri_field()
+    {
         $server_var = $this->config->get('request_uri_field');
         if (!empty($server_var) && (strpos($server_var, 'HTTP_') !== 0 || rcube_utils::check_proxy_whitelist_ip())) {
             return $server_var;
